### PR TITLE
feat(forge): clone verified contracts as a foundry project

### PR DESCRIPTION
**Pull request by Troublor**
This PR is the first step of integrating on-chain contract tweak feature into foundry, as mentioned in #7575. 

We introduce `forge clone` command, which takes the address of an on-chain verified contract and creates a corresponding foundry project in the local filesystem. 

The foundry configuration `foundry.toml` is well generated so that the project is compilation-ready and guaranteed to generate exactly the same code as the on-chain contract. 

We also generate a metadata file, `.clone.meta`, which contains the original address of the contract, constructor arguments, storage layout, etc, which may be useful for downstream tasks.

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great and actually super simple to follow,
I still would like a few more docs, testing looks good

File: crates/forge/bin/cmd/clone.rs
Comment (Troublor) [reply crates/forge/bin/cmd/clone.rs]: I use `#[cfg_attr(test, automock)` here, but `EtherscanClient` trait itself cannot be moved to test mod since it is used in the `run_with_client` function. Please see if this is appropriate or not.

File: crates/config/src/lib.rs
Comment (ZhangZhuoSJTU) [inline crates/config/src/lib.rs:784]: The original code contains a subtle bug. Specifically, let's assume that the current working directory (e.g.,`/A/B/C`), and `self.root`(e.g.,`/X/Y/Z`), are different. Moreover, we further assume `self.src = "src"`. The bug occurs when there is a valid directory at `/A/B/C/src`, because the function `sources(&self.src)` calls `utils::canonicalized(&self.src)`. Consequently, the `sources` returned will be `/A/B/C/src`, although the expected result is `src` or `/X/Y/Z/src`.

**Comment by ZhangZhuoSJTU** [comment]
Hey @mattsse! Big thanks for checking out our PR 😊 

I was actually wondering, do you think it would be a good idea to update the Foundry book now, or should we wait until after the PR is merged? Cheers for the advice!

File: crates/forge/bin/cmd/clone.rs
Comment (mattsse) [inline crates/forge/bin/cmd/clone.rs:320]: could you open an issue for this with a link to the docs, we likely just miss bindings for this setting

**Review by mattsse — CHANGES_REQUESTED** [review]
mind giving this another look @klkvr 

I think this is looking good, I'd love to see one `forgetest!` integration test for this

File: crates/forge/bin/cmd/clone.rs
Comment (ZhangZhuoSJTU) [reply crates/forge/bin/cmd/clone.rs:320]: Sure thing! I've opened an [issue](https://github.com/foundry-rs/compilers/issues/99). Since this seems directly related to the `compilers` repo, I do not report it in the `foundry` repo.

File: crates/forge/bin/cmd/clone.rs
Comment (DaniPopes) [inline crates/forge/bin/cmd/clone.rs:290]: You should be able to use [`Metadata::Settings`](https://docs.rs/foundry-block-explorers/latest/foundry_block_explorers/contract/struct.Metadata.html#method.settings) then assign the fields back into the config, instead of manually going through each field.

File: crates/config/src/lib.rs
Comment (ZhangZhuoSJTU) [reply crates/config/src/lib.rs:784]: @DaniPopes @mattsse if you have a moment, could you kindly review these changes? Given that this code might affect other components of Foundry, your insights would be greatly appreciated.

**Comment by klkvr** [comment]
This looks great, still reading through, but tested and it works correctly.

One thing I've noticed is that foundry.toml ends up with a lot of remappings in some cases, can we perhaps write them to remappings.txt instead?

**Comment by mds1** [comment]
Small UX comment: Should this be `forge init --clone <address>` instead? Might be nice to take multiple addresses there too

Also could be nice to unify this with `cast etherscan-source` in some way, since they feel familiar

**Comment by ZhangZhuoSJTU** [comment]
> One thing I've noticed is that foundry.toml ends up with a lot of remappings in some cases, can we perhaps write them to remappings.txt instead?

@klkvr Absolutely! We'll move these mappings to mapping.txt in later commits.

**Comment by ZhangZhuoSJTU** [comment]
> Small UX comment: Should this be `forge init --clone <address>` instead? Might be nice to take multiple addresses there too
>
> Also could be nice to unify this with `cast etherscan-source` in some way, since they feel familiar

@mds1 Changing to `forge init --clone <address>` sounds like a great idea. However, supporting the cloning of multiple on-chain contracts into the same directory may be challenging because these contracts may have different compilation optimization flags. Any suggestions or ideas?

We are also thinking about unifying `forge clone` and `cast etherscan-source`, but how to proceed remains unknown. My initial idea is to completely rewrite `cast etherscan-source`, which, however, seems dangerous given the potential downstream applications of `cast etherscan-source`. Another idea is to add a new option to `cast etherscan-source`, which also seems inelegant. Any suggestions on this matter?

File: crates/forge/bin/cmd/clone.rs
Comment (ZhangZhuoSJTU) [reply crates/forge/bin/cmd/clone.rs:290]: Hi! Thanks for the suggestion! It's greatly appreciated. 

Currently, we utilize `Metadata::settings()` to retrieve the values for each individual field. I'm thinking to implementing a function that serves as the opposite of `Config::solc_settings`. Do you think that would be an appropriate addition?

File: crates/forge/bin/cmd/clone.rs
Comment (ZhangZhuoSJTU) [reply crates/forge/bin/cmd/clone.rs:318]: Wait for the release of `foundry-compilers` 0.3.16 to include this update.

File: crates/forge/bin/cmd/clone.rs
Comment (ZhangZhuoSJTU) [reply crates/forge/bin/cmd/clone.rs]: Fixed.

File: crates/forge/bin/cmd/clone.rs
Comment (ZhangZhuoSJTU) [reply crates/forge/bin/cmd/clone.rs:340]: @DaniPopes A second thought. Using settings to update config seems to be a unique use case of `forge clone` and possibly out of scope for a generic function for Config. Might it be advisable to stick with our current approach?

**Comment by ZhangZhuoSJTU** [comment]
I've been considering an update to UX, i.e., changing the command from `forge clone` to `forge init --clone <address>`. Would love to hear thoughts and suggestions on whether this is a step in the right direction. @mattsse @DaniPopes @Evalir

**Comment by mattsse** [comment]
keeping forge clone makes sense to me

File: crates/forge/bin/cmd/clone.rs
Comment (mattsse) [inline crates/forge/bin/cmd/clone.rs:81]: nit: move this impl block below the struct def

File: crates/forge/bin/cmd/clone.rs
Comment (mattsse) [inline crates/forge/bin/cmd/clone.rs:109]: use tokio::time::sleep here because we're in async land

File: crates/forge/bin/cmd/clone.rs
Comment (mattsse) [inline crates/forge/bin/cmd/clone.rs:132]: style nit
```suggestion
    pub(crate) async fn collect_metadata_from_client<C: EtherscanClient>(
        address: Address,
        client: &C,
    ) -> Result<Metadata> {
```

File: crates/forge/bin/cmd/clone.rs
Comment (mattsse) [reply crates/forge/bin/cmd/clone.rs:369]: this should be available now

File: crates/forge/bin/cmd/clone.rs
Comment (mattsse) [inline crates/forge/bin/cmd/clone.rs:507]: can we move this into the tests module?

**Review by mattsse — CHANGES_REQUESTED** [review]
great!

last pedantic nits

File: crates/forge/bin/cmd/clone.rs
Comment (Troublor) [reply crates/forge/bin/cmd/clone.rs:507]: This trait is used by the function `run_with_client`, so we cannot move this trait into the test module because otherwise there will be `trait not found` compilation errors in the cargo build. The rationale of this trait is to allow mocking for Etherscan clients. I have made my best to use `#[cfg(test)]` and `#[cft_attr(test, automock)]` so that the mocking is only compiled in `test` build target.

This is the recommended practice in the doc of `mockall`: https://github.com/asomers/mockall?tab=readme-ov-file#usage

**Comment by Troublor** [comment]
@mattsse Correct me if I am wrong. I think the latest patch https://github.com/foundry-rs/compilers/pull/100 has not yet been released in a new version of `foundry-compilers`. The latest version of `foundry-compilers` is `0.3.15` while the commit of the patch is not included. 

Since `foundry` declares `foundry-compilers` as a dependency with version `^0.3.14`, the `simple_counter_for_loop_unchecked_increment` field is still not available.

**Comment by Troublor** [comment]
@mattsse The `forgetest!` integrated test for `forge clone` seems to suffer from Etherscan rate limit. Since CIs are executed in parallel, sleeping in the test case does not avoid the rate limit. Do you have any suggestions?

File: crates/config/src/lib.rs
Comment (Troublor) [inline crates/config/src/lib.rs:2837]: Our PR modifies the function `config.project_paths()`: https://github.com/foundry-rs/foundry/pull/7576/commits/09aada8415ed531f8375d2af2b42fc8bf873c9b3#diff-0f348074359029c2930eb72b52f482c2e0f4495dc61b95bc919f155a8d2fd82c. 
The paths in the config (e.g., test path in this test case) will always be converted to a `PathBuf` relative to the foundry project root. So the correct assertion in this test case should be `./mytest`, where `.` denotes the current foundry project root.

**Comment by Troublor** [comment]
Thanks for your effort @mattsse! We have future addressed some issues in tests. Could you trigger the CI test and see if this PR works?

**Comment by mattsse** [comment]
ty @Troublor 

I'll give this a final review tmrw

**Comment by ZhangZhuoSJTU** [comment]
There's an unusual CI test failure here, although the CI appears correctly in [our repo](https://github.com/EtherDebug/foundry/pull/13).

**Comment by mattsse** [comment]
I'm unable to push directly to this branch for some reason,
could you please include my last commit from https://github.com/foundry-rs/foundry/compare/master...mattsse:foundry:matt/forge-clone?expand=1

then this is g2m

**Review by mattsse — APPROVED** [review]

**Comment by Troublor** [comment]
@mattsse I just merged the latest commit on master branch. Sorry to interrupt the CI that you just triggered. Could you please trigger again?

**Comment by mattsse** [comment]
tysm for this!

**Pull request merged**
